src/zenml/post_execution/base_view.py

```diff
@@ -0,0 +1,142 @@
+
+
+class BaseView(ABC):
+    """Base class for post-execution views.
+
+    Subclasses should override the following attributes/methods:
+    - `MODEL_CLASS` should be set to the model class the subclass is wrapping.
+    - The `model` property should return `self._model` cast to the correct type.
+    - (Optionally) `REPR_KEYS` can be set to include a list of custom attributes
+        in the `__repr__` method.
+    """
+
+    MODEL_CLASS = BaseResponseModel  # The model class to wrap.
+    REPR_KEYS = ["id"]  # Keys to include in the `__repr__` method.
+
+    def __init__(self, model: BaseResponseModel):
+        """Initializes a view for the given model.
+
+        Args:
+            model: The model to create a view for.
+        """
+        # Check that the model is of the correct type.
+        if not isinstance(model, self.MODEL_CLASS):
+            raise TypeError(
+                f"Model of {self.__class__.__name__} must be of type "
+                f"{self.MODEL_CLASS.__name__} but is {type(model).__name__}."
+            )
+
+        # Validate that all `REPR_KEYS` are valid.
+        for key in self.REPR_KEYS:
+            if (
+                key not in model.__fields__
+                and key not in self._custom_view_properties
+            ):
+                raise ValueError(
+                    f"Key {key} in {self.__class__.__name__}.REPR_KEYS is "
+                    f"neither a field of {self.MODEL_CLASS.__name__} nor a "
+                    f"custom property of {self.__class__.__name__}."
+                )
+
+        self._model = model
+
+    @property
+    def _custom_view_properties(self) -> List[str]:
+        """Returns a list of custom view properties.
+
+        Returns:
+            A list of custom view properties.
+        """
+        return [attr for attr in dir(self) if not attr.startswith("__")]
+
+    @property
+    @abstractmethod
+    def model(self) -> BaseResponseModel:
+        """Returns the underlying model.
+
+        Subclasses should override this property to return `self._model` with
+        the correct model type.
+
+        E.g. `return cast(ArtifactResponseModel, self._model)`
+
+        Returns:
+            The underlying model.
+        """
+
+    def __getattribute__(self, __name: str) -> Any:
+        """Returns the attribute with the given name.
+
+        This method is overridden so we can access the model fields as if they
+        were attributes of this class.
+
+        Args:
+            __name: The name of the attribute to return.
+
+        Returns:
+            The attribute with the given name.
+        """
+        # Handle special cases that are required by the `dir` call below
+        if __name in ("__dict__", "__class__"):
+            return super().__getattribute__(__name)
+
+        # Check the custom view properties first in case of overwrites
+        if __name in [attr for attr in dir(self) if not attr.startswith("__")]:
```

Comment:
Could be a set instead of list which is better for lookups

Reply:
Sure, done.